generator/osm_source.cpp

```diff
@@ -248,47 +333,70 @@ class MainFeaturesEmitter
       m_world.reset(new TWorldGenerator(info));
   }
 
-  void operator()(FeatureBuilder1 fb)
+  void operator()(FeatureBuilder1 & fb) override
   {
-    uint32_t const coastType = Type(NATURAL_COASTLINE);
-    bool const hasCoast = fb.HasType(coastType);
+    static uint32_t const placeType = classif().GetTypeByPath({"place"});
+    uint32_t const type = fb.GetParams().FindType(placeType, 1);
 
-    if (m_coasts)
+    auto hotelIndex = numeric_limits<size_t>::max();
+
+    if (type != ftype::GetEmptyValue() && !fb.GetName().empty())
     {
-      if (hasCoast)
-      {
-        CHECK(fb.GetGeomType() != feature::GEOM_POINT, ());
-        // leave only coastline type
-        fb.SetType(coastType);
-        (*m_coasts)(fb);
-      }
-      return;
+      m_places.ReplaceEqualInRect(
+          Place(fb, type),
+          [](Place const & p1, Place const & p2) { return p1.IsEqual(p2); },
+          [](Place const & p1, Place const & p2) { return p1.IsBetterThan(p2); });
     }
-
-    if (hasCoast)
+    else if ((hotelIndex = m_bookingDataset.GetMatchingHotelIndex(fb)) !=
+             numeric_limits<size_t>::max())
     {
-      fb.PopExactType(Type(NATURAL_LAND));
-      fb.PopExactType(coastType);
+      m_skippedElements << DebugPrint(fb.GetMostGenericOsmId()) << endl;
+
+      // Make a hotel a simple building.
+      if (fb.GetGeomType() == feature::GEOM_AREA)
+      {
+        // Remove all information about a hotel.
+        auto params = fb.GetParams();
+        params.ClearName();
+        auto & meta = params.GetMetadata();
+        meta.Drop(feature::Metadata::EType::FMD_STARS);
+        meta.Drop(feature::Metadata::EType::FMD_WEBSITE);
+        meta.Drop(feature::Metadata::EType::FMD_PHONE_NUMBER);
+
+        auto & types = params.m_Types;
+        types.erase(remove_if(begin(types), end(types), [](uint32_t type)
+        {
+          static auto const & c = classif();
+          static auto tourism = c.GetTypeByPath({"tourism"});
```

Comment:
mgsergio <notifications@github.com> writes:

> In generator/osm_source.cpp:
>
>> +      // Make a hotel a simple building.
>> +      if (fb.GetGeomType() == feature::GEOM_AREA)
>> +      {
>> +        // Remove all information about a hotel.
>> +        auto params = fb.GetParams();
>> +        params.ClearName();
>> +        auto & meta = params.GetMetadata();
>> +        meta.Drop(feature::Metadata::EType::FMD_STARS);
>> +        meta.Drop(feature::Metadata::EType::FMD_WEBSITE);
>> +        meta.Drop(feature::Metadata::EType::FMD_PHONE_NUMBER);
>> +
>> +        auto & types = params.m_Types;
>> +        types.erase(remove_if(begin(types), end(types), [](uint32_t type)
>> +        {
>> +          static auto const & c = classif();
>> +          static auto tourism = c.GetTypeByPath({"tourism"});
>
> What's the point?

The point is in static objects initialization. They're initialized
correctly, therefore there're hidden synchronization guards around
static objects. You don't need them here, plain const is enough here.

>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

-- 
Best regards, Yuri